dart/collision/CollisionDetector.h

```diff
@@ -43,169 +43,179 @@
   ///
-  /// The contact force acting on bodyNode2 is -force, which is the opposite
-  /// direction of the force.
-  Eigen::Vector3d force;
+  /// Internally, this function creates shared_ptr from unique_ptr returned from
+  /// createCollisionGroup() so the performance would be slighly worse than
+  /// using std::make_unique.
+  std::shared_ptr<CollisionGroup> createCollisionGroupAsSharedPtr();
 
-  /// First colliding body node
-  dynamics::WeakBodyNodePtr bodyNode1;
+  /// Create a collision group from any objects that are supported by
+  /// CollisionGroup::addShapeFramesOf().
+  ///
+  /// The objects can be any of ShapeFrame, std::vector<ShapeFrame>,
+  /// CollisionGroup, BodyNode, and Skeleton.
+  ///
+  /// Note that this function adds only the ShapeFrames of each object at the
+  /// moment of this function is called. The aftwerward changes of the objects
+  /// does not affect on this CollisionGroup.
+  template <typename... Args>
+  std::unique_ptr<CollisionGroup> createCollisionGroup(const Args&... args);
 
-  /// Second colliding body node
-  dynamics::WeakBodyNodePtr bodyNode2;
+  /// Helper function that creates and returns CollisionGroup as shared_ptr.
+  template <typename... Args>
+  std::shared_ptr<CollisionGroup> createCollisionGroupAsSharedPtr(
+      const Args&... args);
 
-  /// First colliding shape of the first body node
-  dynamics::ShapePtr shape1;
+  /// Perform collision detection for group.
+  virtual bool collide(CollisionGroup* group,
+                       const CollisionOption& option, CollisionResult& result) = 0;
 
-  /// Second colliding shape of the first body node
-  dynamics::ShapePtr shape2;
+  /// Perform collision detection for group1-group2.
+  virtual bool collide(CollisionGroup* group1, CollisionGroup* group2,
+                       const CollisionOption& option, CollisionResult& result) = 0;
 
-  /// Penetration depth
-  double penetrationDepth;
+protected:
 
-  // TODO(JS): triID1 will be deprecated when we don't use fcl_mesh
-  /// \brief
-  int triID1;
+  class CollisionObjectManager;
+  class NoneSharingCollisionObjectManager;
+  class SharingCollisionObjectManager;
 
-  // TODO(JS): triID2 will be deprecated when we don't use fcl_mesh
-  /// \brief
-  int triID2;
+  /// Constructor
+  CollisionDetector() = default;
 
-  // TODO(JS): userData is an experimental variable.
-  /// \brief User data.
-  void* userData;
-};
+  /// Claim CollisionObject associated with shapeFrame. New CollisionObject
+  /// will be created if it hasn't created yet for shapeFrame.
+  std::shared_ptr<CollisionObject> claimCollisionObject(
+      const dynamics::ShapeFrame* shapeFrame);
 
-/// \brief class CollisionDetector
-class CollisionDetector
-{
-public:
-  /// \brief Constructor
-  CollisionDetector();
+  /// Create CollisionObject
+  virtual std::unique_ptr<CollisionObject> createCollisionObject(
+      const dynamics::ShapeFrame* shapeFrame) = 0;
```

Comment:
Is there a particular reason that we create these as a ``unique_ptr`` when it seems as though we always end up converting them into a ``shared_ptr``? Is this to make it impossible to jumble up ``claimCollisionObject(~)`` with ``createCollisionObject(~)``?